lib/mutgen.c

```diff
@@ -187,93 +184,348 @@ mutation_model_check_validity(mutation_model_t *self)
+mutation_matrix_model_alloc(mutation_model_t *self, size_t num_alleles, char **alleles,
     double *root_distribution, double *transition_matrix)
 {
     int ret = 0;
+    mutation_matrix_t *params = &self->params.mutation_matrix;
 
     memset(self, 0, sizeof(mutation_model_t));
     if (num_alleles < 2) {
         ret = MSP_ERR_INSUFFICIENT_ALLELES;
         goto out;
     }
-    self->alleles = calloc(num_alleles, sizeof(*self->alleles));
-    self->root_distribution = malloc(num_alleles * sizeof(*self->root_distribution));
-    self->transition_matrix
-        = malloc(num_alleles * num_alleles * sizeof(*self->transition_matrix));
-    if (self->alleles == NULL || self->root_distribution == NULL
-        || self->transition_matrix == NULL) {
+    params->num_alleles = num_alleles;
+    params->alleles = calloc(num_alleles, sizeof(*params->alleles));
+    params->allele_length = calloc(num_alleles, sizeof(*params->allele_length));
+    params->root_distribution = malloc(num_alleles * sizeof(*params->root_distribution));
+    params->transition_matrix
+        = malloc(num_alleles * num_alleles * sizeof(*params->transition_matrix));
+    if (params->alleles == NULL || params->allele_length == NULL
+        || params->root_distribution == NULL || params->transition_matrix == NULL) {
         ret = MSP_ERR_NO_MEMORY;
         goto out;
     }
-    self->num_alleles = num_alleles;
-    memcpy(self->root_distribution, root_distribution,
+    memcpy(params->root_distribution, root_distribution,
         num_alleles * sizeof(*root_distribution));
-    memcpy(self->transition_matrix, transition_matrix,
+    memcpy(params->transition_matrix, transition_matrix,
         num_alleles * num_alleles * sizeof(*transition_matrix));
-    ret = mutation_model_copy_alleles(self, alleles);
+    ret = mutation_matrix_copy_alleles(params, alleles);
+    if (ret != 0) {
+        goto out;
+    }
+    self->choose_root_state = &mutation_matrix_choose_root_state;
+    self->transition = &mutation_matrix_transition;
+    self->print_state = &mutation_matrix_print_state;
+    self->free = &mutation_matrix_free;
+
+    ret = mutation_matrix_check_validity(params);
     if (ret != 0) {
         goto out;
     }
-    ret = mutation_model_check_validity(self);
+out:
+    return ret;
+}
+
+int MSP_WARN_UNUSED
+slim_mutator_alloc(mutation_model_t *self, int32_t mutation_type_id,
+        int64_t next_mutation_id, size_t block_size)
+{
+    int ret = 0;
+    slim_mutator_t *params = &self->params.slim_mutator;
+    slim_mutation_metadata_t *slim_md = NULL;
+
+    memset(self, 0, sizeof(mutation_model_t));
+    
+    /* In practice this is the minimum we can support */
+    block_size = GSL_MAX(block_size, 128);
+    ret = tsk_blkalloc_init(&params->allocator, block_size);
+    if (ret != 0) {
+        ret = msp_set_tsk_error(ret);
+        goto out;
+    }
+
+    slim_md = malloc(sizeof(*slim_md));
```

Comment:
I'm not sure there's much point in storing this in the struct - since it's a fixed size, it would be simpler/more obvious what's going on if we alloced it statically in the function that's using it.